Sources/StreamChat/Controllers/ChannelController/LivestreamChannelController.swift

```diff
@@ -0,0 +1,894 @@
+
+        case .greaterThan, .greaterThanOrEqual:
+            messages.insert(contentsOf: newMessages, at: 0)
+
+        case .around, .none:
+            messages = newMessages
+        }
+    }
+
+    // MARK: - EventsControllerDelegate
+
+    public func eventsController(_ controller: EventsController, didReceiveEvent event: Event) {
+        guard let channelEvent = event as? ChannelSpecificEvent, channelEvent.cid == cid else {
+            return
+        }
+
+        callback { [weak self] in
+            self?.handleChannelEvent(event)
+        }
+    }
+
+    // MARK: - Helpers
+
+    /// Helper method to execute the callbacks on the main thread.
+    func callback(_ action: @escaping () -> Void) {
+        DispatchQueue.main.async {
+            action()
+        }
+    }
+
+    private func handleChannelEvent(_ event: Event) {
+        switch event {
+        case let messageNewEvent as MessageNewEvent:
+            handleNewMessage(messageNewEvent.message)
+
+        case let localMessageNewEvent as NewMessagePendingEvent:
+            handleNewMessage(localMessageNewEvent.message)
+
+        case let messageUpdatedEvent as MessageUpdatedEvent:
+            handleUpdatedMessage(messageUpdatedEvent.message)
+
+        case let messageDeletedEvent as MessageDeletedEvent:
+            if messageDeletedEvent.isHardDelete {
+                handleDeletedMessage(messageDeletedEvent.message)
+                return
+            }
+            handleUpdatedMessage(messageDeletedEvent.message)
+
+        case let newMessageErrorEvent as NewMessageErrorEvent:
+            guard let message = messages.first(where: { $0.id == newMessageErrorEvent.messageId }) else {
+                return
+            }
+            let errorMessage = message.changing(state: .sendingFailed)
+            handleUpdatedMessage(errorMessage)
+
+        case let reactionNewEvent as ReactionNewEvent:
+            handleNewReaction(reactionNewEvent)
+
+        case let reactionUpdatedEvent as ReactionUpdatedEvent:
+            handleUpdatedReaction(reactionUpdatedEvent)
+
+        case let reactionDeletedEvent as ReactionDeletedEvent:
+            handleDeletedReaction(reactionDeletedEvent)
+
+        case let channelUpdatedEvent as ChannelUpdatedEvent:
+            handleChannelUpdated(channelUpdatedEvent)
+
+        default:
+            break
+        }
+    }
+
+    private func handleNewMessage(_ message: ChatMessage) {
+        var currentMessages = messages
+
+        // If message already exists, update it instead
+        if currentMessages.contains(where: { $0.id == message.id }) {
+            handleUpdatedMessage(message)
+            return
+        }
+
+        // If we don't have the first page loaded, do not insert new messages
+        // they will be inserted once we load the first page again.
+        if !hasLoadedAllNextMessages {
+            return
+        }
+
+        currentMessages.insert(message, at: 0)
+        messages = currentMessages
```

Comment:
this is not really thread safe

Reply:
It is, since the events controller always calls the delegate on the main thread. So we only manipulate the state on the main thread

Reply:
in some cases also the contains operation

Reply:
Having the message processing in another queue, will make things more complex. For now, doing this on the main thread should give us enough performance. Later if we think that we need even better performance than we can change this